### DISCUSSION TITLE
Successful build on WSL (Ubuntu 22.04 LTS)

### DISCUSSION THREAD

**Discussion by DJtheRedstoner** [dreamworksanimation/openmoonray#90] [General]
I was able to successfully build and run MoonRay using WSL (Ubuntu). Thanks to everyone who has tried building MoonRay and posted their advice/patches, especially #45.

It’s important to note that my WSL environment was not clean when I started this, so my instructions may not work exactly. Additionally, I’ve upgraded to Ubuntu 22.04 LTS, which may be different than your WSL environment. These instructions should also work for a non-WSL Ubuntu environment.

Here’s the complete patch I used to get things to compile: https://gist.github.com/DJtheRedstoner/55e8d66226d04f9158bf22ccd9bd176f. I didn’t need to make any changes to the source code, only the build files.

Here are some rough instructions to reproduce my build. I recommend fully reading them and all my notes before attempting this, since it’s important to get all the details correct.
```bash
# Install Package Dependencies
sudo add-apt-repository ppa:deadsnakes/ppa # see the notes below for reasoning
sudo apt-get install python3.9 python3.9-dev python3.9-distutils
# You may need other dependencies, such as gcc-9 and Qt

# Clone MoonRay
cd /home/ubuntu # must be this exact directory, see notes
git clone --recurse-submodules https://github.com/dreamworksanimation/openmoonray.git

# Apply Patch
wget https://gist.githubusercontent.com/DJtheRedstoner/55e8d66226d04f9158bf22ccd9bd176f/raw/2d6e6b1f0a97fb9b5438efedd9bf23ab6ab18754/wsl-build.diff
git apply wsl-build.diff

# Setup Environment Variables (IMPORTANT!)
source env.sh

# Build Dependencies
mkdir -p installs/bin # this is where the dependencies will be installed
mkdir build && cd build # build directory for dependencies
cmake ../building
cmake --build . -- -j12 # replace the value with the number of threads you have
cd ..

# Build MoonRay
cmake --preset container-release -DMOONRAY_USE_CUDA=NO # OptiX doesn't work on WSL, see notes
cmake --build --preset container-release -- -j12 # replace the value with the number of threads you have

# Install MoonRay
cmake --install mrbuild --prefix release

# Run a Render
# you'll need to download and extract the test scenes: https://docs.openmoonray.org/getting-started/test-scenes/
cd /mnt/d/Downloads/pbrt_scenes/bathroom
LD_LIBRARY_PATH=/home/ubuntu/openmoonray/release/lib/:/home/ubuntu/openmoonray/installs/lib/ /home/ubuntu/openmoonray/release/bin/moonray_gui -dso_path /home/ubuntu/openmoonray/release/rdl2dso -in scene.rdla -in scene.rdlb
```
Notes:
1. Directories
    MoonRay’s build is based around absolute directories. My patch changes these directories to be within `/home/ubuntu/openmoonray`. If your path is different, you will need to make changes based on my patch.

    | Directory | Purpose |
    | --- | --- |
    | `/home/ubuntu/openmoonray` | Project root, Git root |
    | `/home/ubuntu/openmoonray/build` | Dependency project build root |
    | `/home/ubuntu/openmoonray/installs` | Dependency install location | 
    | `/home/ubuntu/openmoonray/release` | MoonRay release install location |

1. Python Version
The version of boost MoonRay uses will fail to build with Python 3.10 (which Ubuntu 22.04 ships). Because of this, we add the deadsnakes PPA, install Python 3.9 and configure Boost to build using it (`python-config.jam`).

1. GCC Version
    MoonRay doesn’t appear to build on GCC 11 (the default for me), changing to GCC 9 works and is as easy as setting `CC=/usr/bin/gcc-9 CXX=/usr/bin/g++-9` (done by my `env.sh`).

1. Dual ABI Configuration
    MoonRay is designed to be built with the dual ABI macro `_GLIBCXX_USE_CXX11_ABI` set to `0`, however it defaults to `1` with Ubuntu’s packages. Other people have successfully built MoonRay but changing this value to `1` within MoonRay, but I decided to set it to `0` for everything in order to stay closer to the [VFX Reference Platform](https://vfxplatform.com/). Setting this can be done using environment variables: `CXX_FLAGS=-D_GLIBCXX_USE_CXX11_ABI=0` (again done by my `env.sh`), with two additional changes to `building/CMakeLists.txt` because Boost doesn’t use CMake and OpenColorIO overwrites this environment variable for one of its projects. I’m not an expert on C linking and such, so I’m not sure of the benefits and consequences of each approach. This will likely change in the future ones MoonRay migrates to the 2023 VFX RefPlat (which uses `_GLIBCXX_USE_CXX11_ABI=1`).

1. Cyclic Linking Dependencies
    It seems like certain MoonRay shared objects have cyclic dependencies, which the linker I’m using wasn’t too happy about. I’ve fixed these using CMake’s [`LINK_GROUP` generator expression](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:LINK_GROUP). If someone with more experience could shed some light on why only some linkers care and if there’s a better fix, I’d appreciate it.

1. Missing Link Dependency
    `arras/arras4_core/arras4_client/lib/client/api/CMakeLists.txt` is missing a link dependency to `Boost::filesystem`, which causes linking to fail on my seemingly stricter linker. `Boost::filesystem` is on the link path due to transitive dependencies, but not in the correct order. I’m 90% sure this could be considered a bug, if someone can confirm this I’ll make an issue.

1. GPU Support
    I was able to get GPU support to compile & run (after bypassing the `/proc/driver/nvidia/version` checks that don’t work on WSL), but OptiX is seemingly non-functional in WSL:
    - https://forums.developer.nvidia.com/t/optix-on-wsl-windows-subsystem-for-linux/140297
    - https://forums.developer.nvidia.com/t/problem-running-optix-7-6-in-wsl/239355

Once again, thank you to everyone who documented their previous efforts and thanks to the DWA team for open sourcing this project and helping out the community. If anyone is not able (or is able) to reproduce my build like this, please let me know in the comments. I’ll try to help out as best as I can.

And finally, some renders:
Rendered in 24 minutes on my i5-10400
![sample1](https://user-images.githubusercontent.com/52044242/230224286-2915ebb6-e9df-43a4-a5a0-74b6eaee890d.png)
Denoised with OIDN
![sample2](https://user-images.githubusercontent.com/52044242/230224340-7a05e3d3-15f4-452f-957f-866483c801a8.png)

**Comment by Brithub77** [comment]
I tried just that and am now having this issue:

Error: Unknown option "--ignore-preprocessor-errors".

How can I fix it? Is there a way to get around it?

**Reply by DJtheRedstoner** [reply]
What step are you at when you get this error? Is there any other output? That seems like an ISPC argument, but nothing in moonray uses it afaict. Also, I realized I didn't generate the diff correctly, I've updated the gist with a corrected version. If your attempting to follow this procedure, you should probably apply the new patch otherwise you might run into the same link errors as me.

**Reply by gonsolo** [reply]
I added this option because my version of ISPC is newer. You can just remove this from the Makefiles.

**Reply by Brithub77** [reply]
Okay I did as you suggested and am now getting this:

/home/ubuntu/openmoonray/moonray/moonray/cmd/denoise_cmd/main.cc:4:10: fatal error: mcrt_denoise/denoiser/Denoiser.h: No such file or directory
    4 | #include <mcrt_denoise/denoiser/Denoiser.h>

**Reply by Brithub77** [reply]
Never mind about that. I solved this issue by copying mcrt_denoise to my system "include" directory. However, I am still receiving errors:

/usr/bin/ld: CMakeFiles/denoise.dir/main.cc.o: in function `main.cold':
main.cc:(.text.unlikely+0xf3c): undefined reference to `moonray::denoiser::Denoiser::~Denoiser()'
/usr/bin/ld: CMakeFiles/denoise.dir/main.cc.o: in function `main':
main.cc:(.text.startup+0x65f): undefined reference to `moonray::denoiser::Denoiser::Denoiser(moonray::denoiser::DenoiserMode, int, int, bool, bool, std::string*)'
/usr/bin/ld: main.cc:(.text.startup+0x6e8): undefined reference to `moonray::denoiser::Denoiser::denoise(float const*, float const*, float const*, float*, std::string*)'
/usr/bin/ld: main.cc:(.text.startup+0x793): undefined reference to `moonray::denoiser::Denoiser::~Denoiser()'
collect2: error: ld returned 1 exit status
gmake[2]: *** [moonray/moonray/cmd/denoise_cmd/CMakeFiles/denoise.dir/build.make:110: moonray/moonray/cmd/denoise_cmd/denoise] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:12850: moonray/moonray/cmd/denoise_cmd/CMakeFiles/denoise.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2

**Reply by DJtheRedstoner** [reply]
> Never mind about that. I solved this issue by copying mcrt_denoise to my system "include" directory.

This shouldn't be needed (and is probably a bad idea).

Something is definitely wrong with your setup, the include dirs use symlinks so maybe something is wrong with those. Did you clone the repo under a `/mnt` path? This definitely won't work, the WSL <-> Windows filesystem adapter doesn't understand some features (and is very slow!).

**Reply by Brithub77** [reply]
No, I cloned it in /home/ubuntu.

**Comment by masahi** [comment]
@DJtheRedstoner Got it to build on native Ubuntu 22.04 including optix support following this guide. Thank you very much!!
